### PR TITLE
ClusterTemplate drop Enabled flag, create Cluster using ClusterTemplateRevision only 

### PR DESCRIPTION
**Pull request by prachidamle**
Following changes are added:
- drop Enabled flag on ClusterTemplate:
[https://github.com/rancher/rancher/issues/21063, https://github.com/rancher/rancher/issues/21062]

- create Cluster using ClusterTemplateRevision only, create cluster API does not require ClusterTemplateId:  [https://github.com/rancher/rancher/issues/21131, https://github.com/rancher/rancher/issues/21130]

- Minor fixes to correct error message and panic due to null ClusterTemplateRevision.ClusterConfig
[https://github.com/rancher/rancher/issues/21329, https://github.com/rancher/rancher/issues/21327]

Depends on types PR: https://github.com/rancher/types/pull/896

https://github.com/rancher/rancher/issues/21063
https://github.com/rancher/rancher/issues/21062
https://github.com/rancher/rancher/issues/21131
https://github.com/rancher/rancher/issues/21130
https://github.com/rancher/rancher/issues/21329
https://github.com/rancher/rancher/issues/21327

### CODE REVIEW

File: pkg/api/customization/cluster/validator.go [hunk 1]
```diff
@@ -112,7 +112,7 @@ func (v *Validator) validateEnforcement(request *types.APIContext, data map[stri
 	}
 
 	if !enforced {
-		return httperror.NewFieldAPIError(httperror.PermissionDenied, "", fmt.Sprintf("%v is not a valid clusterTemplate to create a cluster", spec.ClusterTemplateID))
+		return httperror.NewFieldAPIError(httperror.PermissionDenied, "", fmt.Sprintf("%v is not a valid clusterTemplate to create a cluster", spec.ClusterTemplateRevisionID))
```
Comment (alena1108) [inline pkg/api/customization/cluster/validator.go]: nit: we have to change the error message

File: pkg/api/customization/cluster/validator.go [hunk 2]
```diff
@@ -132,13 +132,23 @@ func (v *Validator) checkClusterForEnforcement(spec *mgmtclient.Cluster) bool {
...
+	if err != nil {
+		return false, err
 	}
-	templateName := split[1]
 
-	clusterTemp, err := v.ClusterTemplateLister.Get(namespace.GlobalNamespace, templateName)
+	templateIDStr := clusterTempRev.Spec.ClusterTemplateName
+	split = strings.Split(templateIDStr, ":")
+	if len(split) != 2 {
+		return false, fmt.Errorf("error in splitting clusterTemplate name %v", templateIDStr)
```
Comment (alena1108) [inline pkg/api/customization/cluster/validator.go]: will it be a template revision? then we'd have to change the error message

File: pkg/api/store/cluster/cluster_store.go [hunk 3]
```diff
@@ -228,7 +230,7 @@ func loadDataFromTemplate(clusterTemplateRevision *v3.ClusterTemplateRevision, d
 		}
 		val, err := builder.ConvertSimple(question.Type, answer, builder.Create)
 		if err != nil {
-			return nil, err
+			return nil, httperror.WrapAPIError(err, httperror.ServerError, "Error processing ClusterTemplate answers")
```
Comment (alena1108) [inline pkg/api/store/cluster/cluster_store.go]: nit: in some messages we use camelcase like clusterTemplateRevision, in some uppercase like ClusterTemplate.

File: pkg/api/customization/cluster/validator.go (hunk 2 above)
Comment (prachidamle) [reply pkg/api/customization/cluster/validator.go]: no, it is clusterTemplate

File: pkg/api/store/cluster/cluster_store.go (hunk 3 above)
Comment (prachidamle) [reply pkg/api/store/cluster/cluster_store.go]: changing all to use camelcase

File: pkg/api/store/cluster/cluster_store.go [hunk 4]
```diff
@@ -207,15 +208,16 @@ func (r *Store) Create(apiContext *types.APIContext, schema *types.Schema, data
...
 	dataFromTemplate, err := convert.EncodeToMap(clusterTemplateRevision.Spec.ClusterConfig)
 	if err != nil {
 		return nil, err
 	}
 	dataFromTemplate["name"] = convert.ToString(data["name"])
 	dataFromTemplate["description"] = convert.ToString(data[managementv3.ClusterSpecFieldDisplayName])
-	dataFromTemplate["clusterTemplateId"] = convert.ToString(data[managementv3.ClusterSpecFieldClusterTemplateID])
-	dataFromTemplate["clusterTemplateRevisionId"] = convert.ToString(data[managementv3.ClusterSpecFieldClusterTemplateRevisionID])
+	dataFromTemplate[managementv3.ClusterSpecFieldClusterTemplateID] = ref.Ref(clusterTemplate)
+	dataFromTemplate[managementv3.ClusterSpecFieldClusterTemplateRevisionID] = convert.ToString(data[managementv3.ClusterSpecFieldClusterTemplateRevisionID])
```
Comment (mrajashree) [inline pkg/api/store/cluster/cluster_store.go]: Do we need to check if these fields are present before calling convert.ToString?

Comment (alena1108) [reply pkg/api/store/cluster/cluster_store.go]: @mrajashree I think convert already takes care of checking for nil/empty

Comment (prachidamle) [reply pkg/api/store/cluster/cluster_store.go]: no, this method is called only after checking that the template exists in the data. cluster_store: hasTemplate() and cluster_store::validateTemplateInput validates the data[] input.

**Comment by alena1108** [comment]
LGTM

File: pkg/api/store/cluster/cluster_store.go (hunk 4 above)
Comment (prachidamle) [reply pkg/api/store/cluster/cluster_store.go]: yeah and convert return "" if passed nil to it.

**Pull request merged**